### PR TITLE
Use an emit helper (or Object.assign) for JsxSpreadAttributes

### LINKED ISSUE #7270: React emit should have an '_extends' helper for spread attributes
This is effectively a polyfill for `Object.assign`, and we should consider emitting the specific helper.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
This PR introduces an `__assign` helper that we can leverage for React emit, as well as object spread properties down the road (with some tweaking).

Fixes #7270.

@rbuckton @RyanCavanaugh @mhegazy

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -345,6 +345,15 @@ var __extends = (this && this.__extends) || function (d, b) {
     d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
 };`;
 
+        const assignHelper = `
+var __assign = (this && this.__assign) || Object.assign || function(t) {
+    for (var i = 1, n = arguments.length; i < n; i++) {
+        var s = arguments[i];
+        if (s != null) for (var p in s) if (s.hasOwnProperty(p)) t[p] = s[p];
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: The null test is unnecessary

**Comment by DanielRosenwasser** [comment]
Thanks @basarat :smile:

File: src/compiler/emitter.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: This language is ridonculous.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -345,6 +345,15 @@ var __extends = (this && this.__extends) || function (d, b) {
     d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
 };`;
 
+        const assignHelper = `
+var __assign = (this && this.__assign) || Object.assign || function(t) {
```
Comment (kuon) [inline src/compiler/emitter.ts]: Polyfill I found (and what React used) convert `t` into an object, like:

```
if (t == null) {
throw new TypeError('Object.assign cannot be called with null or undefined');
}
var result = Object(t);
```

I'm unsure if this is necessary with current browsers. But https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#Polyfill does to.

File: src/compiler/types.ts [hunk 3]
```diff
@@ -405,6 +405,7 @@ namespace ts {
         JavaScriptFile =     1 << 27,  // If node was parsed in a JavaScript
         ThisNodeOrAnySubNodesHasError = 1 << 28,  // If this node or any of its children had an error
         HasAggregatedChildData = 1 << 29,  // If we've computed data from children and cached it in this node
+        HasJsxSpreadAttribute = 1 << 30,
```
Comment (RyanCavanaugh) [inline src/compiler/types.ts]: Can you log an issue to address the fact that we're now at the effective flag cap for this enum and should break it apart somehow?

**Comment by RyanCavanaugh** [comment]
:+1: `&&` thank you @gaearon

File: src/compiler/emitter.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: For our uses this should it should be unnecessary anyway.

**Pull request merged**